src/badguy/snail.cpp

```diff
@@ -402,13 +402,18 @@ Snail::grab(MovingObject& object, const Vector& pos, Direction dir_)
 void
 Snail::ungrab(MovingObject& object, Direction dir_)
 {
+  auto player = dynamic_cast<Player*> (&object);
   if (!m_frozen)
   {
     if (dir_ == Direction::UP) {
       be_flat();
     }
     else {
-      m_dir = dir_;
+      if (dir_ != Direction::DOWN) {
+        m_dir = dir_;
```

Comment:
Since `player` is only used in this specific case, it may be better to define it here, so a `dynamic_cast` could be spared in the other cases.